x/qgb/keeper/keeper_valset.go

```diff
@@ -194,3 +204,33 @@ func (k Keeper) GetLatestValsetBeforeNonce(ctx sdk.Context, nonce uint64) (*type
 		fmt.Sprintf("couldn't find valset before nonce %d", nonce),
 	)
 }
+
+func (k Keeper) SetEVMAddress(ctx sdk.Context, valAddress sdk.ValAddress, evmAddress gethcommon.Address) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.GetEVMKey(valAddress), evmAddress.Bytes())
+}
+
+func (k Keeper) GetEVMAddress(ctx sdk.Context, valAddress sdk.ValAddress) (gethcommon.Address, bool) {
+	store := ctx.KVStore(k.storeKey)
+	if !store.Has(types.GetEVMKey(valAddress)) {
+		return gethcommon.Address{}, false
+	}
+	addrBytes := store.Get(types.GetEVMKey(valAddress))
+	return gethcommon.BytesToAddress(addrBytes), true
+}
+
+// IsEVMAddressUnique checks if the provided evm address is globally unique. This
+// includes the defaults we set validators when they initially create a validator
+// before registering
+func (k Keeper) IsEVMAddressUnique(ctx sdk.Context, evmAddress gethcommon.Address) bool {
+	store := ctx.KVStore(k.storeKey)
+	addrBytes := evmAddress.Bytes()
+	iterator := sdk.KVStorePrefixIterator(store, []byte(types.EVMAddress))
+	defer iterator.Close()
+	for ; iterator.Valid(); iterator.Next() {
```

Comment:
Iterating over every EVM address won't be fraud proof friendly. Why does it need to be unique anyway?

Reply:
My understanding is the system is not setup to handle duplicates. See https://github.com/celestiaorg/celestia-app/blob/e2b154d8314872eba6f0d8d5ea18fdb58b201150/x/qgb/types/validator.go#L181-L183

Reply:
we technically don't need it, but the currently implementation relies on maps heavily, so if each is not unique then we could end up overriding a previously added validator. Doing something different would require a dramatic re-write.

Adding this instance of iteration is bad, but we will also need to fix a lot of other instances as well before we can have uncompressed fraud proofs. Do you think its worth optimizing for that now?

Reply:
Say there are two validators A and B with voting power 10 and 100. If they had the same EVM address, a signature from one of them would be impossible to distinguish if it was the one with 10 voting power or 100

Reply:
>Adding this instance of iteration is bad, but we will also need to fix a lot of other instances as well before we can have uncompressed fraud proofs. Do you think its worth optimizing for that now?

No, but just highlighting